stm32-modules/thermocycler-gen2/scripts/calibrate_temperature.py

```diff
@@ -0,0 +1,321 @@
+import numpy as np
+import datetime
+
+from serial.tools.list_ports import grep
+from typing import  Dict, Tuple, List
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Calculate and program thermocycler offset coefficients")
+    parser.add_argument('-t', '--thermocycler', type=str, required=False, default=None,
+                        help='The USB port that the thermocycler is connected to')
+    parser.add_argument('-p', '--probe', type=str, required=True, default=None,
+                        help='The USB port that the multiprobe is connected to')
+    parser.add_argument('-c', '--check', required=False, action='store_true',
+                        help='Run through the temperature settings again after the test to check data')
+    parser.add_argument('-f', '--file', type=argparse.FileType('w'), required=False, default=None,
+                        help='File to write data to. If not specified, a new file will be generated.')
+    return parser.parse_args()
+
+class Multiprobe():
+    """A class to communicate with the Eutechnics 4690 thermometer"""
+    def __init__(self, port: str = None):
+        self.ser = serial.Serial(port, 
+                                 baudrate = 115200,
+                                 timeout=2,
+                                 parity = serial.PARITY_NONE,
+                                 stopbits = serial.STOPBITS_ONE,
+                                 bytesize = serial.EIGHTBITS)
+        # Make sure the probe is connected right
+        self.get_reading(retries=3)
+    
+    @dataclass 
+    class Readings:
+        """Class to encapsulate readings from the multiprobe wells"""
+        h1: float  # Well H1
+        a1: float  # Well A1
+        f4: float  # Well F4
+        c4: float  # Well C4 
+        f9: float  # Well F9 
+        c9: float  # Well C9 
+        h12: float # Well H12
+        a12: float # Well A12
+
+        def left_temp(self) -> float:
+            return average([self.h1, self.a1])
+
+        def center_temp(self) -> float:
+            return average([self.f4, self.c4, self.f9, self.c9])
+
+        def right_temp(self) -> float:
+            return average([self.h12, self.a12])
+
+        def as_list(self) -> List[float]:
+            return [self.h1, self.a1, self.f4, self.c4, self.f9, self.c9, self.h12, self.a12]
+
+        def uniformity(self) -> float:
+            temps = self.as_list()
+            return (max(temps) - min(temps)) / 2
+
+        def average(self) -> float:
+            return average(self.as_list())
+
+        def to_csv(self) -> str:
+            ret = ''
+            for well in self.as_list():
+                ret += f'{well},'
+            return ret
+    
+    def _read_probe_line(self) -> bytes:
+        """
+        Replaces `serial.readline` because readline does not work with the 
+        endline character used by the probe
+        """
+        ret = bytes()
+        char = b' '
+        while char != b'\r':
```

Comment:
Lol at it using \cr but also https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.read_until